native/src/hooks/useLocation.ts

```diff
@@ -0,0 +1,141 @@
+
+type SuccessfulLocationState = { status: 'ready' }
+type UnavailableLocationState = {
+  status: 'unavailable'
+  message: 'noPermission' | 'notAvailable' | 'timeout' | 'loading'
+}
+
+type LocationStateType = SuccessfulLocationState | UnavailableLocationState
+
+export type LocationType = [number, number]
+
+const locationStateOnError = (error: GeolocationError): UnavailableLocationState => {
+  if (error.code === error.PERMISSION_DENIED) {
+    return {
+      status: 'unavailable',
+      message: 'noPermission'
+    }
+  } else if (error.code === error.POSITION_UNAVAILABLE) {
+    return {
+      status: 'unavailable',
+      message: 'notAvailable'
+    }
+  } else {
+    return {
+      status: 'unavailable',
+      message: 'timeout'
+    }
+  }
+}
+
+export type LocationInformationType = {
+  location: LocationType | null
+  locationState: LocationStateType
+  requestAndDetermineLocation: () => Promise<void>
+}
+
+const useLocation = (useSettingsListener = false): LocationInformationType => {
+  const [locationState, setLocationState] = useState<LocationStateType>({
+    status: 'unavailable',
+    message: 'loading'
+  })
+  const [location, setLocation] = useState<LocationType | null>(null)
+
+  const determineLocation = useCallback(() => {
+    Geolocation.getCurrentPosition(
+      (position: GeolocationResponse) => {
+        setLocation([position.coords.longitude, position.coords.latitude])
+        setLocationState({ status: 'ready' })
+      },
+      (error: GeolocationError) => {
+        setLocationState(locationStateOnError(error))
+      },
+      {
+        enableHighAccuracy: true,
+        timeout: 50000,
+        maximumAge: 3600000
+      }
+    )
+  }, [])
+
+  useEffect(() => {
+    checkLocationPermission().then(locationPermissionStatus => {
+      if (locationPermissionStatus === RESULTS.GRANTED) {
+        setLocationState({
+          message: 'loading',
+          status: 'unavailable'
+        })
+        determineLocation()
+      } else {
+        setLocationState({
+          status: 'unavailable',
+          message: 'noPermission'
+        })
+      }
+    })
+  }, [determineLocation])
+
+  const requestAndDetermineLocation = useCallback(async () => {
+    setLocationState({
+      message: 'loading',
+      status: 'unavailable'
+    })
+    const locationPermissionStatus = await checkLocationPermission()
+
+    if (locationPermissionStatus === RESULTS.BLOCKED) {
+      await openSettings()
+      setLocationState({
+        message: 'noPermission',
+        status: 'unavailable'
+      })
+    } else if (locationPermissionStatus === RESULTS.GRANTED) {
+      determineLocation()
+    } else {
+      if ((await requestLocationPermission()) === RESULTS.GRANTED) {
+        determineLocation()
+      } else {
+        setLocationState({
+          message: 'noPermission',
+          status: 'unavailable'
+        })
+      }
+    }
```

Comment:
```suggestion
      if (locationPermissionStatus === RESULTS.GRANTED || (await requestLocationPermission()) === RESULTS.GRANTED) {
      determineLocation()
    } else {
      setLocationState({
        message: 'noPermission',
        status: 'unavailable'
      })
      if (locationPermissionStatus === RESULTS.BLOCKED) {
        await openSettings()
      }
    }
```

Reply:
hmm. but what if the user changes his permission state while using the app? Then we'd assume he has still granted permission here 🤔

Reply:
Isn't this an issue with your code as well?

Reply:
This part was just copy/pasted from Landing.tsx. What issue are you addressing?

Reply:
Hmm I thought you replied to this saying there is some issue with my code. but then nvm...